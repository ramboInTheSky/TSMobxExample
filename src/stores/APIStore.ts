import { observable, action, toJS } from 'mobx'
import axios from 'axios'

import { getErrorCode, toSnakeCase, toCamelCase } from '../utils'

export { observable, action }

export class Store<T> {
	private url: string
	private fresh = false

	@observable protected data: Array<T> = []

	@observable protected detail: T | Partial<T> = {}

	@observable protected count: number = 0

	@observable protected loading: boolean = false

	@observable public error?: string = undefined

	constructor(uri: string) {
		this.url = uri
	}
	// public API
	// public getItems = (): Array<T> => {
	// 	return toJS(this.data) as Array<T>
	// }

	public get items(): Array<T> {
		return toJS(this.data) as Array<T>
	}

	public get item(): T {
		return toJS(this.detail) as T
	}

	// public getItem = (): T => {
	// 	return toJS(this.detail) as T
	// }

	public get errorMessage(): string {
		return this.error as string
	}

	public get isLoading(): boolean {
		return this.loading
	}

	public get isError(): boolean {
		return !!this.error
	}

	public isFresh = () => this.fresh
	// end public API

	protected apiEnded = (e?: any, customMessage?: string) => {
		this.loading = false
		this.error = e ? customMessage || getErrorCode(e) : undefined
		this.fresh = true
		const self = this
		setTimeout(function () {
			self.fresh = false
		})
	}

	@action
	async getList(url?: string, filter?: string) {
		if (!this.url && !url) return
		try {
			this.loading = true
			let res: any
			res = await axios.get(url || this.url + '?filter=' + filter)
			this.apiEnded()
			this.count = res.data && res.data.count ? res.data.count : 0
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			this.count = 0
			return []
		}
	}

	@action
	async getDetail(item: string) {
		try {
			this.loading = true
			let res: any
			res = await axios.get(`${this.url}/${item}`)
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}

	@action
	async save(item: Partial<T>, id?: string, url?: string) {
		try {
			this.loading = true
			let res: any
			if (id) {
				res = await axios.put(`${url || this.url}/${id}`, toSnakeCase(item))
			} else {
				res = await axios.post(url || this.url, toSnakeCase(item))
			}
			this.apiEnded()
			return toCamelCase(res.data)
		} catch (e) {
			this.apiEnded(e)
			return {}
		}
	}
}
