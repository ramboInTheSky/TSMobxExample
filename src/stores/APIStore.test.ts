import React from 'react'

import { Store } from './APIStore'

jest.mock('axios', () => () => ({
	get: jest.fn(
		(url: string) =>
			url === 'fail' ? Promise.reject({ status: 500 }) : Promise.resolve({ status: 200, data: {} }),
	),
}))

describe('<Store />', () => {
	it('renders correctly', () => {
		const store = new Store('/url')
		expect(store).toMatchSnapshot()
	})
	it('sets isError if API issues', async () => {
		const store = new Store('fail')
		await store.getList()
		setTimeout(() => expect(store.isError).toBe(true))
	})
	it('returns data correctly', async () => {
		const store = new Store('/url')
		await store.getList()
		setTimeout(() => expect(store.isError).toBe(false))
	})
})
