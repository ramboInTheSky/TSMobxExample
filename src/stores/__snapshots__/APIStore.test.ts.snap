// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Store /> renders correctly 1`] = `
Store {
  "apiEnded": [Function],
  "fresh": false,
  "isFresh": [Function],
  "url": "/url",
}
`;
